### DISCUSSION TITLE
JWE verification is valid after providing secret key as Buffer with an invalid string | Node.js

### DISCUSSION THREAD

**Discussion by Hleb-Haiduk-Software** [panva/jose#452] [Q&A] [answered]
Hello Dear Community!
Kindly ask you to help with the following.
I'm trying to understand how the JOSE library works for JWE creating/validation in Node.js.

To create a JWE I need a secret key in KeyObject or Uint8Array format.
I have a secret string in my .env file and want to create a secret key based on that.

    const secret = Buffer.alloc(32, process.env.SECRET_KEY);
    const jwe = await new jose.EncryptJWT({ userId: '1234532' })
    .setProtectedHeader({ alg: 'dir', enc: 'A256GCM' })
    .setIssuedAt()
    .setIssuer('urn:example:issuer')
    .setAudience('urn:example:audience')
    .setExpirationTime('2h')
    .encrypt(secret1);

Works fine, we created a JWE.

Then I want to provide an invalid secret, changing my .env key for testing that JWE will be invalid if the secret key is not correct.
  

    const secretWithInvalidString = Buffer.alloc(32, process.env.INVALID_SECRET);
    const { payload, protectedHeader } = await jose.jwtDecrypt(jwt, secretWithInvalidString, {
      issuer: 'urn:example:issuer',
      audience: 'urn:example:audience'
    });

My JWE is still VALID but it shouldn't be as I changed a secret string in .env file.

After comparing two buffers with different strings I can see that they are equal, but why?

    // returns true but why when provided strings to Buffer are different?
       console.log(console.log(secret.equals(secretWithInvalidString)));

P.S. I haven't worked with Buffers too much before :|

THANKS!

**Comment by panva** [comment, marked as answer]
```
> Buffer.alloc(32, 'djsakldjlsakjdasjkdjsalkdjlskajdlksjdsadjklsajdksajldksa')
<Buffer 64 6a 73 61 6b 6c 64 6a 6c 73 61 6b 6a 64 61 73 6a 6b 64 6a 73 61 6c 6b 64 6a 6c 73 6b 61 6a 64>
> Buffer.alloc(32, 'djsakldjlsakjdasjkdjsalkdjlskajdlksjdsadjklsajdksajldksafhjdshfjkdshfjksa')
<Buffer 64 6a 73 61 6b 6c 64 6a 6c 73 61 6b 6a 64 61 73 6a 6b 64 6a 73 61 6c 6b 64 6a 6c 73 6b 61 6a 64>
```

Notice the same buffer comes out.

What you want to do is e.g. 

```js
Buffer.from('059bde05519b5a8098f1007fc8dd28ad223378a997ed3750edfb0895859ee18d', 'hex')
```

**Reply by panva** [reply]
- Buffer.alloc https://nodejs.org/api/buffer.html#static-method-bufferallocsize-fill-encoding
- Buffer.from https://nodejs.org/api/buffer.html#static-method-bufferfromstring-encoding

**Reply by Hleb-Haiduk-Software** [reply]
Thanks for your reply, but it didn't solve the problem.
I used the following way of secret key creation:

`const secret = Buffer.from(mySecretString, 'hex');`

As a result, I receive the following error: **throw new errors_js_1.JWEInvalid('Invalid Content Encryption Key length');**

okay, I checked the following table (https://github.com/panva/jose/issues/210), where the algorithm **enc: 'A256GCM'** requires a key of 256 bit length.

I tried to provide such a key:

`const secret2 = Buffer.from( new String( mySecretString ), 'hex', 256 );`

But still have the **same error message.**
What am I doing wrong? :)
Still can't hack this library:)

**Reply by panva** [reply]
```js
import * as jose from 'jose';

const secret = Buffer.from('059bde05519b5a8098f1007fc8dd28ad223378a997ed3750edfb0895859ee18d', 'hex');
console.log(secret.byteLength);

const jwe = await new jose.EncryptJWT({ userId: '1234532' })
  .setProtectedHeader({ alg: 'dir', enc: 'A256GCM' })
  .setIssuedAt()
  .setIssuer('urn:example:issuer')
  .setAudience('urn:example:audience')
  .setExpirationTime('2h')
  .encrypt(secret);

console.log(jwe);
```

This works. Secret is a 32 **byte** long secret. That is the same as 256 **bits**.

> I used the following way of secret key creation

But your hex string does not represent 32 bytes when decoded from hex. See `secret.byteLength`

> I tried to provide such a key

Generate the secret in the proper length. You're again trying to use a construct which would truncate whatever your secret string is. Just use one in the proper length in the first place. Also, the third argument in this Buffer API overload you use is documented to be in bytes, not bits.

> Still can't hack this library:)

Your issue is not with this library.

**Reply by Hleb-Haiduk-Software** [reply]
I didn't mean any issue with the library, sure it's only my problem.
Thanks for the help.
